File: impl/typescript/src/topic/adapter.ts
/**
 * TopicAdapter -- interface abstracting backend-specific topic operations.
 *
 * Spec reference: 004 spec.md
 *   - FR-T04: TopicAdapter is the central abstraction for topic operations.
 *   - FR-T05: Adapter registry maps BackendKind to TopicAdapter at runtime.
 *   - FR-T22: Publish returns PublishResult.
 *   - FR-T23: ConfirmationMode controls publish behavior.
 *   - FR-T24: Subscribe returns AsyncGenerator of MessageDelivery.
 *   - FR-T27: MaxMessageSize returns maximum payload size.
 *   - FR-T28: EstimatePublishCost returns CostEstimate.
 *
 * Spec reference: 004 contracts/topic-adapter.md
 *   - Full adapter interface contract with capability requirements.
 *   - Read-only adapters MUST return UnsupportedOperation for CreateTopic/Publish.
 *
 * Spec reference: 004 data-model.md
 *   - TopicAdapter, CreateTopicOpts, PublishOpts, SubscribeOpts, TopicMetadata, CostEstimate.
 */

import type { NeuronPrivateKey } from '../keylib/private-key.js';
import type { NeuronPublicKey } from '../keylib/public-key.js';
import type { BackendKind, ConfirmationMode } from './types.js';
import type { TopicRef } from './topic-ref.js';
import type { TopicMessage } from './message.js';
import type { PublishResult } from './publish-result.js';
import type { MessageDelivery } from './message-delivery.js';
import { adapterNotRegistered } from './errors.js';

// ---------------------------------------------------------------------------
// Supporting types
// ---------------------------------------------------------------------------

/**
 * Options for creating a new topic on a backend.
 * FR-T04: CreateTopic requires transport kind, admin key, and optional config.
 */
export interface CreateTopicOpts {
  /** Which backend to create the topic on. FR-T04 */
  readonly transport: BackendKind;

  /** Key that controls the topic (from keylib 002). FR-T04 */
  readonly adminKey: NeuronPrivateKey;

  /** Optional topic description/memo. */
  readonly memo?: string | undefined;

  /** Backend-specific creation options. FR-T15 */
  readonly config?: Record<string, unknown> | undefined;
}

/**
 * Options for publishing a message.
 * FR-T23: ConfirmationMode controls whether to wait for consensus.
 */
export interface PublishOpts {
  /** Confirmation mode: FIRE_AND_FORGET or WAIT_FOR_CONSENSUS. FR-T23 */
  readonly confirmationMode: ConfirmationMode;
}

/**
 * Options for subscribing to a topic.
 * FR-T25: FromSequence enables resumption from a specific backend sequence.
 */
export interface SubscribeOpts {
  /**
   * Resume from this backend sequence number.
   * FR-T25: If set, the adapter MUST backfill messages from this sequence to current.
   * If absent, subscribe from latest.
   */
  readonly fromSequence?: bigint | undefined;
}

/**
 * Topic metadata returned by Resolve().
 * FR-T04: Contains current state of a topic on its backend.
 */
export interface TopicMetadata {
  /** The topic reference. FR-T04 */
  readonly topicRef: TopicRef;

  /** Current latest sequence number on the backend. FR-T04 */
  readonly sequenceNumber: bigint;

  /** Topic creation timestamp (nanoseconds). */
  readonly createdAt: bigint;

  /** Topic admin public key (if available). */
  readonly adminKey?: NeuronPublicKey | undefined;

  /** Topic memo/description. */
  readonly memo?: string | undefined;
}

/**
 * Estimated cost of publishing a message.
 * FR-T28: Amount and unit for per-message cost estimation.
 */
export interface CostEstimate {
  /** Estimated cost amount. FR-T28 */
  readonly amount: bigint;

  /** Cost unit (e.g., "tinybar", "wei", "USD-cents"). FR-T28 */
  readonly unit: string;
}

// ---------------------------------------------------------------------------
// TopicAdapter interface
// ---------------------------------------------------------------------------

/**
 * The central abstraction for backend-specific topic operations.
 *
 * FR-T04: All topic interactions go through this interface. Each supported
 * backend provides a concrete implementation.
 *
 * Capability contract:
 * - All adapters MUST implement subscribe(), resolve(), maxMessageSize(), supportedTransport().
 * - Read-write adapters MUST implement createTopic() and publish().
 * - Read-only adapters MUST return UnsupportedOperation from createTopic() and publish().
 * - All adapters SHOULD implement estimatePublishCost().
 */
export interface TopicAdapter {
  /**
   * Create a new topic on the backend.
   * FR-T04: Returns UnsupportedOperation on read-only backends.
   */
  createTopic(opts: CreateTopicOpts): Promise<TopicRef>;

  /**
   * Publish a signed TopicMessage to a topic.
   * FR-T04, FR-T22, FR-T23: Returns PublishResult based on confirmation mode.
   */
  publish(ref: TopicRef, msg: TopicMessage, opts?: PublishOpts): Promise<PublishResult>;

  /**
   * Subscribe to a topic message stream.
   * FR-T04, FR-T24, FR-T25: Returns async generator of MessageDelivery.
   */
  subscribe(ref: TopicRef, opts?: SubscribeOpts): AsyncGenerator<MessageDelivery>;

  /**
   * Resolve topic metadata from the backend.
   * FR-T04: Returns topic state including sequence number, creation time, admin key.
   */
  resolve(ref: TopicRef): Promise<TopicMetadata>;

  /**
   * Maximum payload size in bytes for this backend.
   * FR-T27: Publish MUST check message size against this limit BEFORE submitting.
   */
  maxMessageSize(): bigint;

  /**
   * Estimate the cost of publishing a message of the given size.
   * FR-T28: Returns CostEstimate or UnsupportedOperation.
   */
  estimatePublishCost(messageSize: bigint): Promise<CostEstimate>;

  /**
   * The transport kind this adapter handles.
   * FR-T01: Used for adapter registry lookup.
   */
  supportedTransport(): BackendKind;
}

// ---------------------------------------------------------------------------
// Adapter Registry
// ---------------------------------------------------------------------------

/**
 * Registry mapping BackendKind to TopicAdapter.
 *
 * FR-T05: Adapters are registered at runtime. Registering a new adapter
 * MUST NOT require changes to the core API.
 *
 * SC-T05: New adapters can be added without modifying existing code.
 */
export interface AdapterRegistry {
  /**
   * Register a TopicAdapter for its supported transport kind.
   * FR-T05: Registration MUST NOT silently replace existing adapters.
   *
   * @param kind - The BackendKind this adapter handles
   * @param adapter - The adapter implementation
   */
  register(kind: BackendKind, adapter: TopicAdapter): void;

  /**
   * Retrieve the registered adapter for a transport kind.
   *
   * @param kind - The BackendKind to look up
   * @returns The registered adapter, or undefined if not found
   */
  get(kind: BackendKind): TopicAdapter | undefined;
}

/**
 * Default in-memory adapter registry.
 *
 * FR-T05: Simple Map-based registry. Thread safety is not a concern
 * in single-threaded JavaScript environments.
 */
export class DefaultAdapterRegistry implements AdapterRegistry {
  private readonly _adapters = new Map<string, TopicAdapter>();

  /**
   * Register a TopicAdapter.
   *
   * FR-T05: Throws if an adapter is already registered for this transport kind.
   *
   * @param kind - BackendKind this adapter handles
   * @param adapter - The adapter implementation
   * @throws TopicError NEURON-TOPIC-006 if adapter already registered
   */
  register(kind: BackendKind, adapter: TopicAdapter): void {
    if (this._adapters.has(kind)) {
      throw adapterNotRegistered(
        `Adapter already registered for transport kind: "${kind}"`,
      );
    }
    this._adapters.set(kind, adapter);
  }

  /**
   * Retrieve a registered adapter.
   *
   * @param kind - The BackendKind to look up
   * @returns The registered adapter, or undefined if not found
   */
  get(kind: BackendKind): TopicAdapter | undefined {
    return this._adapters.get(kind);
  }

  /**
   * Check whether an adapter is registered for a transport kind.
   *
   * @param kind - The BackendKind to check
   * @returns `true` if an adapter is registered
   */
  has(kind: BackendKind): boolean {
    return this._adapters.has(kind);
  }

  /**
   * Return all registered transport kinds.
   *
   * @returns Array of registered BackendKind values
   */
  registeredKinds(): BackendKind[] {
    return Array.from(this._adapters.keys()) as BackendKind[];
  }
}

File: impl/golang/internal/topic/adapter.go
package topic

import (
	"context"
	"errors"
	"fmt"
	"sync"
	"time"
)

// CreateTopicOpts holds configuration for creating a new topic on a backend.
type CreateTopicOpts struct {
	// Transport specifies which backend to create the topic on.
	Transport BackendKind
	// AdminKey is the key authorized to manage the topic.
	// Will be NeuronPrivateKey once keylib integration is complete.
	AdminKey interface{}
	// Memo is an optional human-readable description of the topic.
	Memo string
	// Config holds backend-specific configuration parameters.
	// For HCS: submitKey, autoRenewPeriod, etc.
	// For Kafka: partitions, replicationFactor, anchoringConfig, etc.
	Config map[string]interface{}
}

// FR-T23: Confirmation modes: FIRE_AND_FORGET, WAIT_FOR_CONSENSUS
// PublishOpts configures the behavior of a publish operation.
type PublishOpts struct {
	// ConfirmationMode determines whether publish blocks for consensus.
	ConfirmationMode ConfirmationMode
}

// FR-T25: Subscribe resumption from sequence with dedup support
// SubscribeOpts configures the behavior of a subscribe operation.
type SubscribeOpts struct {
	// FromSequence, if non-nil, specifies the sequence number to start
	// reading from (inclusive). Nil means start from the latest message.
	FromSequence *uint64
	// Dedup, if non-nil, enables replay detection. Messages whose signature
	// has already been seen are silently dropped.
	Dedup *DeduplicationTracker
	// ErrorHandler, if non-nil, is called when a received message fails
	// deserialization or signature validation. This enables logging/monitoring
	// of dropped messages.
	ErrorHandler func(error)
}

// TopicMetadata holds backend-reported metadata about a topic.
type TopicMetadata struct {
	// TopicRef is the reference to the topic on the backend.
	TopicRef TopicRef
	// SequenceNumber is the latest message sequence number on the topic.
	SequenceNumber uint64
	// CreatedAt is the topic creation timestamp in Unix nanoseconds.
	CreatedAt uint64
	// AdminKey is the public key authorized to manage the topic.
	// Will be *NeuronPublicKey once keylib integration is complete.
	AdminKey interface{}
	// Memo is the topic description set at creation time.
	Memo string
}

// CostEstimate represents an estimated cost for a topic operation.
type CostEstimate struct {
	// Amount is the estimated cost in the smallest unit of the currency.
	Amount uint64 `json:"amount"`
	// Unit is the currency unit (e.g., "tinybar", "wei", "usd-cent").
	Unit string `json:"unit"`
}

// FR-T04: TopicAdapter interface (createTopic, publish, subscribe, resolve)
// TopicAdapter is the unified interface for backend-specific topic operations.
// Each adapter implementation wraps a particular transport backend (HCS, Kafka, ERC log, etc.)
// and provides a common API for creating topics, publishing messages, subscribing, and resolving metadata.
type TopicAdapter interface {
	// CreateTopic creates a new topic on the backend with the given configuration.
	CreateTopic(opts CreateTopicOpts) (TopicRef, error)
	// Publish sends a signed TopicMessage to the specified topic.
	Publish(ref TopicRef, msg TopicMessage, opts PublishOpts) (PublishResult, error)
	// Subscribe opens a subscription channel that delivers messages from the topic.
	// The context controls the lifetime of the subscription goroutine; cancelling
	// the context stops message delivery and closes the returned channel.
	Subscribe(ctx context.Context, ref TopicRef, opts SubscribeOpts) (<-chan MessageDelivery, error)
	// Resolve retrieves metadata about the specified topic from the backend.
	Resolve(ref TopicRef) (TopicMetadata, error)
	// FR-T27: maxMessageSize query and pre-publish size check
	// MaxMessageSize returns the maximum payload size in bytes for this backend.
	MaxMessageSize() uint64
	// FR-T28: estimatePublishCost for fee-bearing backends
	// EstimatePublishCost returns an estimated cost for publishing a message of the given size.
	EstimatePublishCost(messageSize uint64) (CostEstimate, error)
	// SupportedTransport returns the BackendKind this adapter handles.
	SupportedTransport() BackendKind
}

// adapterMu protects the adapterRegistry map from concurrent access.
var adapterMu sync.RWMutex

// FR-T13: Independent channel backing (different backends per channel)
// adapterRegistry maps backend kinds to their adapter implementations.
var adapterRegistry = map[BackendKind]TopicAdapter{}

// FR-T05: Support >=2 built-in adapters with runtime registration
// RegisterAdapter registers a TopicAdapter for its supported transport.
// Returns an error if an adapter is already registered for the same transport.
func RegisterAdapter(adapter TopicAdapter) error {
	if adapter == nil {
		return NewTopicError(ErrInvalidConfig, "adapter must not be nil")
	}

	transport := adapter.SupportedTransport()

	adapterMu.Lock()
	defer adapterMu.Unlock()

	if _, exists := adapterRegistry[transport]; exists {
		return NewTopicError(ErrInvalidConfig,
			fmt.Sprintf("adapter already registered for transport: %s", transport))
	}

	adapterRegistry[transport] = adapter
	return nil
}

// GetAdapter returns the registered TopicAdapter for the given transport.
// Returns ErrUnsupportedTransport if no adapter is registered for the transport.
func GetAdapter(transport BackendKind) (TopicAdapter, error) {
	adapterMu.RLock()
	defer adapterMu.RUnlock()

	adapter, ok := adapterRegistry[transport]
	if !ok {
		return nil, NewTopicError(ErrUnsupportedTransport,
			fmt.Sprintf("no adapter registered for transport: %s", transport))
	}
	return adapter, nil
}

// ResetAdapterRegistry clears all registered adapters. Intended for testing only.
func ResetAdapterRegistry() {
	adapterMu.Lock()
	defer adapterMu.Unlock()

	adapterRegistry = map[BackendKind]TopicAdapter{}
}

// FR-T26: Exponential backoff retry (baseDelay x 2^attempt, max 5)
// RetryConfig configures exponential backoff retry behavior.
type RetryConfig struct {
	// BaseDelay is the initial delay between retries. Default: 1s.
	BaseDelay time.Duration
	// MaxRetries is the maximum number of retry attempts. Default: 5.
	MaxRetries int
}

// DefaultRetryConfig returns the default retry configuration.
func DefaultRetryConfig() RetryConfig {
	return RetryConfig{
		BaseDelay:  1 * time.Second,
		MaxRetries: 5,
	}
}

// RetryWithBackoff executes fn with exponential backoff retry.
// Retries on transient errors (TopicError with BackendUnavailable kind).
// Returns immediately on permanent errors or success.
func RetryWithBackoff(config RetryConfig, fn func() error) error {
	if config.BaseDelay <= 0 {
		config.BaseDelay = DefaultRetryConfig().BaseDelay
	}
	if config.MaxRetries <= 0 {
		config.MaxRetries = DefaultRetryConfig().MaxRetries
	}

	var lastErr error
	for attempt := 0; attempt <= config.MaxRetries; attempt++ {
		lastErr = fn()
		if lastErr == nil {
			return nil
		}

		// Only retry on transient (BackendUnavailable) errors.
		var topicErr TopicError
		if errors.As(lastErr, &topicErr) && topicErr.Kind == ErrBackendUnavailable {
			if attempt < config.MaxRetries {
				delay := config.BaseDelay * (1 << uint(attempt))
				time.Sleep(delay)
				continue
			}
		}

		// Permanent error: return immediately.
		return lastErr
	}

	return lastErr
}
